File: my-leetcode-solutions/same-tree.ts
// My solution to 100 - Same Tree

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isSameTree(p: TreeNode | null, q: TreeNode | null): boolean {
    if(p === null && q === null) {
        return true;
    } else if (p === null || q === null) {
        return false;
    }

    if (isSameTree(p.left, q.left) === false) {
        return false;
    } else if (isSameTree(p.right, q.right) === false) {
        return false;
    } else {
        isSameTree(p.left, q.left);
        isSameTree(p.right, q.right);
        return true;
    }
};

// Time Complexity:

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null

    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}

const pTestCases = [
    new TreeNode(1, new TreeNode(2), new TreeNode(3)),
    new TreeNode(1, new TreeNode(2)),
    new TreeNode(1, new TreeNode(2), new TreeNode(1)),
];

const qTestCases = [
    new TreeNode(1, new TreeNode(2), new TreeNode(3)),
    new TreeNode(1, null, new TreeNode(2)),
    new TreeNode(1, new TreeNode(1), new TreeNode(2)),
];

for (let i = 0; i < pTestCases.length; i++) {
    const output = isSameTree(pTestCases[i], qTestCases[i]);
    console.log(`Output: ${output} for p as ${JSON.stringify(pTestCases[i])} and q as ${JSON.stringify(qTestCases[i])}`);
}

export {};

File: my-leetcode-solutions-py/same-tree.py
# My solution to 100 - Same Tree

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSameTree(self, p: TreeNode | None, q: TreeNode | None) -> bool:
        pass


# Time Complexity:

p_test_cases = [
    [1, 2, 3],
    [1, 2],
    [1, 2, 1],
]

q_test_cases = [
    [1, 2, 3],
    [1, None, 2],
    [1, 1, 2],
]

solution = Solution()

for i in range(len(p_test_cases)):
    trees = []

    for values in (p_test_cases[i], q_test_cases[i]):
        nodes = [
            TreeNode(value) if value is not None else None
            for value in values
        ]
        children = iter(nodes[1:])

        for node in nodes:
            if node is not None:
                node.left = next(children, None)
                node.right = next(children, None)

        trees.append(nodes[0])

    p, q = trees
    output = solution.isSameTree(p, q)
    print(
        f"Output: {output!r} for p as {p_test_cases[i]!r} "
        f"and q as {q_test_cases[i]!r}"
    )
